File: model/src/main/java/net/artux/pda/model/quest/story/StoryDataModel.java
package net.artux.pda.model.quest.story;

import net.artux.pda.model.items.ArmorModel;
import net.artux.pda.model.items.ArtifactModel;
import net.artux.pda.model.items.DetectorModel;
import net.artux.pda.model.items.ItemModel;
import net.artux.pda.model.items.ItemType;
import net.artux.pda.model.items.MedicineModel;
import net.artux.pda.model.items.WeaponModel;
import net.artux.pda.model.items.WearableModel;
import net.artux.pda.model.user.Gang;
import net.artux.pda.model.user.GangRelation;

import java.io.Serializable;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;

import lombok.Data;

@Data
public class StoryDataModel implements Serializable {

    private String name;
    private String nickname;
    private String login;
    private String avatar;
    private int money;
    private int xp;
    private int pdaId;
    private Gang gang;
    private GangRelation relations;

    private List<ParameterModel> parameters;
    private List<StoryStateModel> storyStates;

    private List<ArmorModel> armors;
    private List<WeaponModel> weapons;
    private List<MedicineModel> medicines;
    private List<DetectorModel> detectors;
    private List<ArtifactModel> artifacts;
    private List<ItemModel> bullets;

    public String getAvatar() {
        if (avatar.contains("http"))
            return avatar;
        else return "textures/avatars/a" + avatar + ".png";
    }

    public boolean containsCurrent() {
        return getCurrentState() != null;
    }

    public <T extends ItemModel> void addItem(T item) {
        if (item.getType().isCountable()) {
            addAsCountable(item);
        } else {
            item.setQuantity(1);
            addAsNotCountable(item);
        }
    }

    private <T extends ItemModel> void addAsNotCountable(T item) {
        ItemType type = item.getType();
        if (type.isWearable()) {
            boolean userWears = getEquippedWearable(type) != null;
            ((WearableModel) item).setEquipped(!userWears);
        }
        item.setQuantity(1);
        addAsIs(item);
    }

    private <T extends ItemModel> void addAsCountable(T itemEntity) {
        Optional<? extends ItemModel> optionalItem = getAllItems()
                .stream()
                .filter(item -> item.getBaseId() == itemEntity.getBaseId())
                .findFirst();
        if (optionalItem.isPresent()) {
            ItemModel item = optionalItem.get();
            item.setQuantity(item.getQuantity() + itemEntity.getQuantity());
        } else {
            addAsIs(itemEntity);
        }
    }

    private <T extends ItemModel> void addAsIs(T itemEntity) {
        switch (itemEntity.getType()) {
            case BULLET:
                getBullets().add(itemEntity);
                break;
            case ARMOR:
                getArmors().add((ArmorModel) itemEntity);
                break;
            case PISTOL:
            case RIFLE:
                getWeapons().add((WeaponModel) itemEntity);
                break;
            case ARTIFACT:
                getArtifacts().add((ArtifactModel) itemEntity);
                break;
            case DETECTOR:
                getDetectors().add((DetectorModel) itemEntity);
                break;
            case MEDICINE:
                getMedicines().add((MedicineModel) itemEntity);
                break;
        }
    }


    public StoryStateModel getCurrentState() {
        if (storyStates != null)
            for (StoryStateModel state : storyStates) {
                if (state.isCurrent())
                    return state;
            }
        return null;
    }

    public StoryStateModel getStateByStoryId(int id) {
        if (storyStates != null)
            for (StoryStateModel state : storyStates) {
                if (state.getStoryId() == id)
                    return state;
            }
        return null;
    }

    public HashMap<String, Integer> getParametersMap() {
        HashMap<String, Integer> map = new HashMap<>();
        if (parameters != null && parameters.size() > 0)
            for (ParameterModel param : parameters) {
                map.put(param.getKey(), param.getValue());
            }
        return map;
    }

    public WearableModel getEquippedWearable(ItemType type) {
        for (ItemModel item : getAllItems()) {
            if (item instanceof WearableModel)
                if (item.getType() == type)
                    if (((WearableModel) item).isEquipped()) {
                        return (WearableModel) item;
                    }
        }
        return null;
    }

    public void setCurrentWearable(WearableModel itemModel) {
        WearableModel current = getEquippedWearable(itemModel.getType());
        if (current != null)
            current.setEquipped(false);
        if (current != itemModel)
            itemModel.setEquipped(true);
    }

    public ItemModel getItemByBaseId(int baseId) {
        for (ItemModel item : getAllItems()) {
            if (item.getBaseId() == baseId) {
                return item;
            }
        }
        return null;
    }

    public List<ItemModel> getAllItems() {
        List<ItemModel> items = new LinkedList<>();
        items.addAll(weapons);
        items.addAll(armors);
        items.addAll(artifacts);
        items.addAll(medicines);
        items.addAll(detectors);
        items.addAll(this.bullets);

        return items;
    }

    public float getTotalWeight() {
        float weight = 0;
        for (ItemModel item : getAllItems())
            weight += item.getWeight() * item.getQuantity();
        return weight;
    }

    public Rang getRang() {
        return getRang(xp);
    }

    public static Rang getRang(int xp) {
        Rang previousRang = Rang.BEGINNER;
        for (Rang rang : Rang.values()) {
            if (rang.getXp() > xp)
                return previousRang;
            else previousRang = rang;
        }
        return Rang.FINAL;
    }

    public enum Rang {
        BEGINNER(0, 0),
        NEW(1, 1000),
        STALKER(2, 3000),
        EXPERIENCE(3, 6000),
        OLD(4, 1000),
        MASTER(5, 16000),
        FINAL(6, Integer.MAX_VALUE, true);

        private final int id;
        private final int xp;
        private final boolean last;

        Rang(int id, int xp) {
            this(id, xp, false);
        }

        Rang(int id, int xp, boolean last) {
            this.id = id;
            this.xp = xp;
            this.last = last;
        }

        public int getId() {
            return id;
        }

        public boolean isLast() {
            return last;
        }

        public int getXp() {
            return xp;
        }

        public Rang getNextRang() {
            if (id < Rang.values().length - 1)
                return Rang.values()[id + 1];
            else return null;
        }
    }

}

File: model/src/main/java/net/artux/pda/model/quest/story/StoryDataModel.kt
package net.artux.pda.model.quest.story

import net.artux.pda.model.items.ArmorModel
import net.artux.pda.model.items.ArtifactModel
import net.artux.pda.model.items.DetectorModel
import net.artux.pda.model.items.ItemModel
import net.artux.pda.model.items.ItemType
import net.artux.pda.model.items.MedicineModel
import net.artux.pda.model.items.WeaponModel
import net.artux.pda.model.items.WearableModel
import net.artux.pda.model.user.Gang
import net.artux.pda.model.user.GangRelation
import java.io.Serializable
import java.util.LinkedList
import java.util.Optional

data class StoryDataModel(
    var name: String? = null,
    var nickname: String? = null,
    var login: String,
    var money: Int = 0,
    var xp: Int = 0,
    var pdaId: Int = 0,
    var gang: Gang? = null,
    var relations: GangRelation? = null
) : Serializable {

    var avatar: String? = null
        get() = if (field!!.contains("http")) field else "textures/avatars/a$field.png"
    var parameters: MutableList<ParameterModel> = mutableListOf()
    var storyStates: MutableList<StoryStateModel> = mutableListOf()
    var armors: MutableList<ArmorModel> = mutableListOf()
    var weapons: MutableList<WeaponModel> = mutableListOf()
    var medicines: MutableList<MedicineModel> = mutableListOf()
    var detectors: MutableList<DetectorModel> = mutableListOf()
    var artifacts: MutableList<ArtifactModel> = mutableListOf()
    var bullets: MutableList<ItemModel> = mutableListOf()
    var items: MutableList<ItemModel> = mutableListOf()

    fun containsCurrent(): Boolean {
        return currentState != null
    }

    fun <T : ItemModel> addItem(item: T) {
        if (item.type.isCountable) {
            addAsCountable(item)
        } else {
            item.quantity = 1
            addAsNotCountable(item)
        }
    }

    private fun <T : ItemModel> addAsNotCountable(item: T) {
        val type = item.type
        if (type.isWearable) {
            val userWears = getEquippedWearable(type) != null
            (item as WearableModel).isEquipped = !userWears
        }
        item.quantity = 1
        addAsIs<T>(item)
    }

    private fun <T : ItemModel> addAsCountable(itemEntity: T) {
        val optionalItem: Optional<out ItemModel> = allItems
            .stream()
            .filter { item: ItemModel -> item.baseId == itemEntity.baseId }
            .findFirst()
        if (optionalItem.isPresent) {
            val item = optionalItem.get()
            item.quantity = item.quantity + itemEntity.quantity
        } else {
            addAsIs(itemEntity)
        }
    }

    private fun <T : ItemModel> addAsIs(itemEntity: T) {
        when (itemEntity.type) {
            ItemType.BULLET -> bullets.add(itemEntity)
            ItemType.ARMOR -> armors.add(itemEntity as ArmorModel)
            ItemType.PISTOL, ItemType.RIFLE -> weapons.add(itemEntity as WeaponModel)
            ItemType.ARTIFACT -> artifacts.add(itemEntity as ArtifactModel)
            ItemType.DETECTOR -> detectors.add(itemEntity as DetectorModel)
            ItemType.MEDICINE -> medicines.add(itemEntity as MedicineModel)
            else -> {}
        }
    }

    val currentState: StoryStateModel?
        get() {
            for (state in storyStates) {
                if (state.current) return state
            }
            return null
        }

    fun getStateByStoryId(id: Int): StoryStateModel? {
        for (state in storyStates) {
            if (state.storyId == id) return state
        }
        return null
    }

    val parametersMap: HashMap<String, Int>
        get() {
            val map = HashMap<String, Int>()
            if (parameters.size > 0) for (param in parameters) {
                map[param.key] = param.value
            }
            return map
        }

    fun getEquippedWearable(type: ItemType): WearableModel? {
        for (item in allItems) {
            if (item is WearableModel) if (item.type === type) if (item.isEquipped && item.quantity > 0) {
                return item
            }
        }
        return null
    }

    fun setCurrentWearable(itemModel: WearableModel?) {
        if (itemModel != null) {
            val current = getEquippedWearable(itemModel.type)
            if (current != null) current.isEquipped = false

            if (current !== itemModel) itemModel.isEquipped = true
        }
    }

    fun getItemByBaseId(baseId: Int): ItemModel? {
        for (item in allItems) {
            if (item.baseId == baseId) {
                return item
            }
        }
        return null
    }

    val allItems: List<ItemModel>
        get() {
            val items: MutableList<ItemModel> = LinkedList()
            items.addAll(weapons)
            items.addAll(armors)
            items.addAll(artifacts)
            items.addAll(medicines)
            items.addAll(detectors)
            items.addAll(bullets)
            items.addAll(this.items)
            return items
        }
    val totalWeight: Float
        get() {
            var weight = 0f
            for (item in allItems) weight += item.weight * item.quantity
            return weight
        }

    val rang: Rang
        get() = getRang(xp)

    enum class Rang(
        val id: Int,
        val xp: Int,
        val isLast: Boolean = false
    ) {
        BEGINNER(0, 0), NEW(1, 1000), STALKER(2, 3000), EXPERIENCE(3, 6000), OLD(
            4,
            10000
        ),
        MASTER(5, 16000), FINAL(6, Int.MAX_VALUE, true);

        val nextRang: Rang?
            get() = if (id < values().size - 1) values()[id + 1] else null
    }

    companion object {
        @JvmStatic
        fun getRang(xp: Int): Rang {
            var previousRang = Rang.BEGINNER
            for (rang in Rang.values()) {
                previousRang = if (rang.xp > xp) return previousRang else rang
            }
            return Rang.FINAL
        }

    }
}